File: liberty/server/compress.py
import sys, os
import struct

def FindBestChar(data):
	#do histogram on the data
	CharCount = [0]*256

	for i in data:
		CharCount[ord(i)] += 1

	#find an entry to use
	UsefulEntry = [0, len(data)]
	for i in xrange(0, len(CharCount)):
		if CharCount[i] < UsefulEntry[1]:
			UsefulEntry = [i, CharCount[i]]

	return UsefulEntry

data = open(sys.argv[1], "r").read()
OrigData = data

#########################
#layer 1 compression
#########################
print "Layer 1"
Layer1Size = len(data)

#print "Using char %02x, found %d times" % (UsefulEntry[0], UsefulEntry[1])
CompressKey = chr(FindBestChar(data)[0])

#now start looking for all the data that duplicates, if found then replace it with our special byte
#need at least 4 to be useful otherwise ignore it as the format is the following
#Key Count DupeByte
#if count is 0 then just output the key byte itself as it is supposed to exist

#first byte is our key to decompress with
OutData = CompressKey
i = 0
while(i < len(data)):
	if i < (len(data) - 3):
		#if 4 match then count how many to do
		if (data[i] == data[i+1]) and (data[i] == data[i+2]) and (data[i] == data[i+3]):
			MatchCount = 4
			while((i + MatchCount < len(data)) and (MatchCount < 258) and (data[i] == data[i+MatchCount])):
				MatchCount += 1

			#output our special flag, length, and char. length is 1 to 255 translating to 4 to 259
			OutData += CompressKey + chr(MatchCount - 3) + data[i]
			i += (MatchCount - 1)	#due to +1 in the loop
		elif data[i] == CompressKey:
			#found our compress key, handle it
			OutData += CompressKey + '\x00'
		else:
			OutData += data[i]

	elif data[i] == CompressKey:
		#found our compress key, handle it
		OutData += CompressKey + '\x00'
	else:
		OutData += data[i]
	i += 1

print "Layer1: %d -> %d" % (Layer1Size, len(OutData))

#########################
#layer 2 compression
#########################
print "Layer 2"
Layer2Size = len(OutData)

#add our starting header, 0 indicates count of 0
data = chr(0) + OutData

#we did try running the following code in a loop but the end compression was worse

CurData = ""

UsefulEntry = FindBestChar(data);
print "Using char %02x, found %d times" % (UsefulEntry[0], UsefulEntry[1])
CompressKey = chr(FindBestChar(data)[0])

DataList = []
CurData = CompressKey

#replace all entries of the compress key with key + chr(255)
data = data.replace(CompressKey, CompressKey + chr(255))

print "Data len", len(data)

for i in xrange(255, 2, -1):
	#if too many entries then stop
	if len(DataList) == 256:
		break

	pos = 0
	while(pos < len(data)):
		#if too many entries then stop
		if len(DataList) == 256:
			break

		#if we hit the data entry that matches our search byte then insert a dummy entry to avoid messing up data
		if len(DataList) == ord(CompressKey):
			DataList += [CompressKey + CompressKey]

		#if we find an entry then find/replace it all
		if (data.find(data[pos:pos+i], pos+i) != -1) and (data[pos:pos+i] not in DataList):
			#mass find/replace
			DataList += [data[pos:pos+i]]
			data = data.replace(data[pos:pos+i], CompressKey + chr(len(DataList) - 1))

		#next position
		pos += 1

#now create the entries for curdata
for i in xrange(0, len(DataList)):
		CurData += chr(len(DataList[i])) + DataList[i]

#add the modified version
SmallestData = CurData + data

print "Layer2: %d -> %d" % (Layer2Size, len(SmallestData))

#########################
#layer 3 compression
#########################
print "Layer 3"
Layer3Size = len(SmallestData)

#get our special character
SpecialKey = FindBestChar(SmallestData)
if SpecialKey[1] != 0:
	print "Unable to compress, all byte values used"
	sys.exit(0)

SpecialKey = chr(SpecialKey[0])

#add our starting header
data = SpecialKey + SmallestData

#start compressing, we are looking for 2 byte pairs to replace
SmallestData = data + " "
CurData = data
while(len(CurData) < len(SmallestData)):
	#make sure we know what our smallest is then re-generate on it
	SmallestData = CurData
	CurData = ""

	#do histogram on the data
	CharCount = [0]*256

	for i in SmallestData:
		CharCount[ord(i)] += 1

	#find an entry to use
	UsefulEntry = [0, len(SmallestData)]
	for i in xrange(0, len(CharCount)):
		#not allowed to use the special key
		if i == ord(SpecialKey):
			continue

		if CharCount[i] < UsefulEntry[1]:
			UsefulEntry = [i, CharCount[i]]

	#print "Using char %02x, found %d times" % (UsefulEntry[0], UsefulEntry[1])
	CompressKey = chr(UsefulEntry[0])

	DataList = []
	CurData = CompressKey

	#make a copy as we will be modifying it
	data = SmallestData

	#make a histogram of 2 byte combos
	WordCount = [0]*0x10000
	for i in xrange(0, len(SmallestData) - 1):
		Entry = (ord(SmallestData[i+1]) << 8) | ord(SmallestData[i])
		WordCount[Entry] += 1

	#find the most used
	BestMatch = [0, 0]
	for i in xrange(0, len(WordCount)):
		if WordCount[i] > BestMatch[1]:
			BestMatch = [i, WordCount[i]]

	#we need at least 3 entries to be useful. 3 bytes in header + 1 per replacement. with 2 replacements = 5 bytes while original is only 4
	if BestMatch[1] >= 3:
		RepChars = chr(BestMatch[0] & 0xff) + chr(BestMatch[0] >> 8)
		CurData = CompressKey + RepChars
		i = 0
		while(i < len(data)):
			if ord(data[i]) == ord(CompressKey):
				#found our compress key, leave it alone
				CurData += CompressKey + SpecialKey
			elif (i < (len(data) - 2)) and (ord(data[i]) == (BestMatch[0] & 0xff)) and (ord(data[i+1]) == ((BestMatch[0] >> 8) & 0xff)) and (SpecialKey != data[i+2]):
				#found an entry to replace
				CurData += CompressKey
				i += 1
			else:
				CurData += data[i]
			i += 1;
	else:
		CurData = data

#add our special key, when found again we stop this loop and it is used to indicate bytes to not replace
OutData = SpecialKey + SmallestData
print "Layer3: %d -> %d" % (Layer3Size, len(OutData))

#delete the original if it exists
try:
	os.unlink(sys.argv[i] + ".a")
except:
	pass

OutData = struct.pack("<H", len(OrigData)) + OutData
open(sys.argv[1] + ".a", "w").write(OutData)
print "%d -> %d, %0.2f%% compressed" % (len(OrigData), len(OutData), (1.0 - (float(len(OutData)) / float(len(OrigData)))) * 100.0)

File: liberty/server/compress.c
#include <malloc.h>
#include <stdio.h>
#include <string.h>
#include <fcntl.h>

typedef struct BestCharStruct
{
	unsigned char Char;
	unsigned int Count;
} BestCharStruct;

typedef struct OutDataStruct
{
	unsigned char *Data;
	unsigned int DataLen;
	unsigned int DataMaxLen;
} OutDataStruct;

void FreeOutData(OutDataStruct *OutData)
{
	if(!OutData->Data)
		free(OutData->Data);
	memset(OutData, 0, sizeof(OutDataStruct));
}

void AddToOutData(OutDataStruct *OutData, void *InData, unsigned int len)
{
	unsigned char *Data = (char *)InData;

	//add data to the buffer
	if(!OutData->Data)
	{
		OutData->Data = (char *)malloc(len + 1024);
		OutData->DataLen = 0;
		OutData->DataMaxLen = len + 1024;
	}
	else if((OutData->DataLen + len) >= OutData->DataMaxLen)
	{
		OutData->DataMaxLen = OutData->DataLen + len + 1024;
		OutData->Data = (char *)realloc(OutData->Data, OutData->DataMaxLen);
	}

	memcpy(&OutData->Data[OutData->DataLen], Data, len);
	OutData->DataLen += len;
}

BestCharStruct FindBestChar(unsigned char *data, unsigned int len, unsigned int SkipChar)
{
	unsigned int CharCount[256];
	unsigned int i;
	BestCharStruct CurBest;

	memset(CharCount, 0, sizeof(CharCount));

	//loop through all entries getting counts
	for(i = 0; i < len; i++)
		CharCount[data[i]]++;

	//find the lowest count
	CurBest.Char = 0;
	CurBest.Count = len;
	for(i = 0; i < 256; i++)
	{
		//if we have a skip char then skip it
		if((SkipChar < 256) && (i == SkipChar))
			continue;

		if(CharCount[i] < CurBest.Count)
		{
			CurBest.Char = i & 0xff;
			CurBest.Count = CharCount[i];
		}
	};
	return CurBest;
}

OutDataStruct Phase1(OutDataStruct *InData)
{
	unsigned char CompressKey;
	unsigned int i;
	unsigned int Count;
	OutDataStruct OutData;
	unsigned char *Data;
	unsigned int len;

	//start phase 1
	Data = InData->Data;
	len = InData->DataLen;
	CompressKey = FindBestChar(Data, len, 0x100).Char;

	memset(&OutData, 0, sizeof(OutData));
	AddToOutData(&OutData, &CompressKey, 1);

	i = 0;
	while(i < len)
	{
		//plenty of data to work on, if we find 4 matching entries in a row then we can compress
		//as our setup is CompressByte MatchCount OrigByte so we have at least 1 byte
		if((i < (len - 3)) && (Data[i] == Data[i+1]) && (Data[i] == Data[i+2]) && (Data[i] == Data[i+3]))
		{
			//find out how many we have, maximum being 258 as value of 1 translates to 4 bytes
			Count = 4;
			while(((i + Count) < len) && (Count < 258) && (Data[i] == Data[i+Count]))
				Count++;

			//output the special values
			Count -= 3;
			AddToOutData(&OutData, &CompressKey, 1);
			AddToOutData(&OutData, &Count, 1);
			AddToOutData(&OutData, &Data[i], 1);
			i += (Count + 2);
		}
		else if(Data[i] == CompressKey)
		{
			//found our compress key (highly unlikely), add an indicator so we decompress properly
			AddToOutData(&OutData, &CompressKey, 1);
			AddToOutData(&OutData, "\x00", 1);
		}
		else
			AddToOutData(&OutData, &Data[i], 1);

		i++;
	}

	return OutData;
}

char *memfind(char *InBuffer, unsigned long InBufLen, char *FindData, unsigned long FindLen)
{
	unsigned char *FindPos;
	unsigned long CurPos;

	//start searching for matches
	CurPos = 0;
	while(CurPos < InBufLen)
	{
		//quickly find the first byte we are looking for
		FindPos = memchr(&InBuffer[CurPos], FindData[0], InBufLen - CurPos);

		//if no match or not enough room then fail
		if((!FindPos) || (((unsigned long)FindPos + FindLen) > (unsigned long)&InBuffer[InBufLen]))
			break;

		//have a position, try it
		if(memcmp(FindPos, FindData, FindLen) == 0)
			return FindPos;

		//not there, try again
		CurPos = (unsigned long)FindPos - (unsigned long)InBuffer;
		CurPos++;
	};

	//failed
	return (char *)0;
}

OutDataStruct Phase2(OutDataStruct *InData)
{
	unsigned char CompressKey;
	OutDataStruct OutData;
	OutDataStruct CurData;
	OutDataStruct TempData;
	OutDataStruct DataList[256];
	unsigned int DataListLen;
	unsigned int i;
	unsigned int Pos;
	long x;
	unsigned int Found;
	unsigned char *DataPos;
	unsigned char *LastDataPos;

	memset(&OutData, 0, sizeof(OutData));
	memset(&CurData, 0, sizeof(CurData));
	memset(&TempData, 0, sizeof(TempData));
	DataListLen = 0;
	memset(DataList, 0, sizeof(DataList));

	//setup the initial buffer by prepending a null byte to it to indicate done
	AddToOutData(&OutData, "\x00", 1);
	AddToOutData(&OutData, InData->Data, InData->DataLen);

	//go find a compression key to use
	CompressKey = FindBestChar(OutData.Data, OutData.DataLen, 0x100).Char;
	printf("compress key: %02x\n", CompressKey);

	//create a copy of the data, replacing all entries of the compress key with compress key + 0xff
	for(i = 0; i < OutData.DataLen; i++)
	{
		AddToOutData(&CurData, &OutData.Data[i], 1);
		if(CurData.Data[i] == CompressKey)
			AddToOutData(&CurData, "\xff", 1);
	}

	//get rid of our temporary out data
	FreeOutData(&OutData);

	//start creating a new data entry
	AddToOutData(&OutData, &CompressKey, 1);

	//CurData contains the data to process on
	for(i = 255; i > 2; i--)
	{
		//go try to find a match
		Pos = 0;
		while((Pos + i) < CurData.DataLen)
		{
			//see if we can find a match in the buffer
			DataPos = memfind(&CurData.Data[Pos + i], CurData.DataLen - Pos - i, &CurData.Data[Pos], i);
			if(DataPos)
			{

				//add it to our list, find all entries and replace

				//if our data list matches our compress key then we need to insert a dummy entry
				if(DataListLen == CompressKey)
				{
					AddToOutData(&DataList[DataListLen], &CompressKey, 1);
					AddToOutData(&DataList[DataListLen], &CompressKey, 1);
					DataListLen++;
				}

				//add our copy of the data
				AddToOutData(&DataList[DataListLen], &CurData.Data[Pos], i);

				//do full replacements in the data
				FreeOutData(&TempData);
				AddToOutData(&TempData, CurData.Data, Pos);

				AddToOutData(&TempData, &CompressKey, 1);
				AddToOutData(&TempData, &DataListLen, 1);

				//DataPos is the 2nd position found, mark the end of the first finding so we can copy the data
				//inbetween when the loop starts
				LastDataPos = &CurData.Data[Pos + i];
				x = 0;
				while(DataPos)
				{
					//add everything between the new position and the last one
					AddToOutData(&TempData, LastDataPos, (unsigned long)DataPos - (unsigned long)LastDataPos);

					//add special marker
					AddToOutData(&TempData, &CompressKey, 1);
					AddToOutData(&TempData, &DataListLen, 1);

					//adjust last position to start copying from
					LastDataPos = DataPos + i;

					//calculate new amount of data to search in
					x = (long)&CurData.Data[CurData.DataLen] - (long)LastDataPos;
					if((unsigned long)LastDataPos >= (unsigned long)&CurData.Data[CurData.DataLen])
						break;

					//find new entry
					DataPos = memfind(LastDataPos, x, &CurData.Data[Pos], i);
				};

				//add rest of the data
				if((unsigned long)LastDataPos < (unsigned long)&CurData.Data[CurData.DataLen])
					AddToOutData(&TempData, LastDataPos, (unsigned long)&CurData.Data[CurData.DataLen] - (unsigned long)LastDataPos);

				DataListLen++;

				//if we hit 256 entries then bail
				if(DataListLen == 256)
					break;

				//swap buffers
				FreeOutData(&CurData);
				CurData = TempData;
				memset(&TempData, 0, sizeof(TempData));
			}
			else
				Pos++;

			//if we hit 256 entries then bail
			if(DataListLen == 256)
				break;
		}
	}

	//add all of the entries
	for(i = 0; i < DataListLen; i++)
	{
		AddToOutData(&OutData, &DataList[i].DataLen, 1);
		AddToOutData(&OutData, DataList[i].Data, DataList[i].DataLen);
	}

	printf("Key: %02x, %d entries\n", CompressKey, DataListLen);

	AddToOutData(&OutData, CurData.Data, CurData.DataLen);

	//all done
	return OutData;
}

OutDataStruct Phase3(OutDataStruct *InData)
{
	OutDataStruct OutData;
	OutDataStruct CurData;
	OutDataStruct SmallestData;
	BestCharStruct SpecialKey;
	BestCharStruct CompressKey;
	unsigned short BestMatchShort;
	unsigned int BestMatchCount;
	unsigned int WordCount[0x10000];
	unsigned int i;

	memset(&OutData, 0, sizeof(OutData));
	memset(&CurData, 0, sizeof(CurData));
	memset(&SmallestData, 0, sizeof(SmallestData));

	SpecialKey = FindBestChar(InData->Data, InData->DataLen, 0x100);
	if(SpecialKey.Count != 0)
	{
		puts("Failed to find valid char for phase 3\n");
		return OutData;
	}

	//add the special key in then the normal data
	AddToOutData(&CurData, &SpecialKey.Char, 1);
	AddToOutData(&CurData, InData->Data, InData->DataLen);

	//loop until we have a buffer that isn't smaller
	SmallestData.DataLen = CurData.DataLen + 1;
	while(CurData.DataLen < SmallestData.DataLen)
	{
		//setup smallest to be the current buffer and rest current
		FreeOutData(&SmallestData);
		SmallestData = CurData;
		memset(&CurData, 0, sizeof(CurData));

		//find a char to use
		CompressKey = FindBestChar(SmallestData.Data, SmallestData.DataLen, (unsigned int)SpecialKey.Char);

		//do histogram
		memset(WordCount, 0, sizeof(WordCount));
		for(i = 0; i < SmallestData.DataLen - 1; i++)
			WordCount[*(unsigned short *)&SmallestData.Data[i]]++;

		//go find the worst offender
		BestMatchShort = 0;
		BestMatchCount = 0;
		for(i = 0; i < 0x10000; i++)
		{
			if(WordCount[i] > BestMatchCount)
			{
				BestMatchShort = (unsigned short)i;
				BestMatchCount = WordCount[i];
			}
		}

		//if it is big enough then start replacing
		if(BestMatchCount >= 3)
		{
			//add the char and two bytes it replaces
			AddToOutData(&CurData, &CompressKey.Char, 1);
			AddToOutData(&CurData, &BestMatchShort, 2);

			//cycle through all data
			for(i = 0; i < SmallestData.DataLen; i++)
			{
				//if we found the compress key leave it alone
				if(SmallestData.Data[i] == CompressKey.Char)
				{
					AddToOutData(&CurData, &CompressKey.Char, 1);
					AddToOutData(&CurData, &SpecialKey.Char, 1);
				}
				else if((i < (SmallestData.DataLen - 2)) &&
					(*(unsigned short *)&SmallestData.Data[i] == BestMatchShort) &&
					(SmallestData.Data[i+2] != SpecialKey.Char))
				{
					//found an entry to replace
					AddToOutData(&CurData, &CompressKey, 1);
					i++;
				}
				else
					AddToOutData(&CurData, &SmallestData.Data[i], 1);
			}
		}
		else
			break;	//nothing valid to replace, stop looking
	}

	//add our special key to the beginning
	FreeOutData(&CurData);
	AddToOutData(&CurData, &SpecialKey.Char, 1);
	AddToOutData(&CurData, SmallestData.Data, SmallestData.DataLen);

	return CurData;
}

unsigned char * Compress(unsigned char *Data, unsigned int *len)
{
	OutDataStruct OutData;
	unsigned int PrevLen;

	memset(&OutData, 0, sizeof(OutData));

	OutData.Data = Data;
	OutData.DataLen = *len;
	PrevLen = *len;
	OutData = Phase1(&OutData);
	printf("Phase1: %d -> %d\n", PrevLen, OutData.DataLen);

	PrevLen = OutData.DataLen;
	OutData = Phase2(&OutData);
	printf("Phase2: %d -> %d\n", PrevLen, OutData.DataLen);


	PrevLen = OutData.DataLen;
	OutData = Phase3(&OutData);
	printf("Phase3: %d -> %d\n", PrevLen, OutData.DataLen);

	*len = OutData.DataLen;
	return OutData.Data;
}
/*
int main(int argc, char **argv)
{
	int fd;
	unsigned char *Buffer;
	OutDataStruct CompressBuffer;
	int FileLen;
	unsigned char NewName[256];

	//read in
	fd = open(argv[1], O_RDONLY);
	FileLen = lseek(fd, 0, SEEK_END);
	lseek(fd, 0, SEEK_SET);
	Buffer = malloc(FileLen);
	read(fd, Buffer, FileLen);
	close(fd);

	//compress
	CompressBuffer = Compress(Buffer, FileLen);

	//write out
	sprintf(NewName, "%s.z", argv[1]);
	unlink(NewName);
	fd = creat(NewName, 0777);
	write(fd, &FileLen, 2);
	write(fd, CompressBuffer.Data, CompressBuffer.DataLen);
	close(fd);
}
*/
